File: CodingSample/DotNet/Enterprise/Networking/TcpServerTest/DemoApp1/Models/ShopModel.cs
using System.Net.Sockets;

namespace DemoApp.Models;

public class ShopModel(string server)
{
    public async Task<ItemInfo> ReadItemInfoAsync(string name)
    {
        //Step 1
        var connection = new TcpClient(server, 4000);
        //Step 2
        var remote = connection.GetStream();
        using var reader = new StreamReader(remote);
        using var writer = new StreamWriter(remote) { AutoFlush = true };
        await reader.ReadLineAsync(); //read greeting message
        await writer.WriteLineAsync(name);
        string message = await reader.ReadLineAsync();
        //Step 3
        connection.Close();
        if(message != null)
            return ItemInfo.Parse(message);
        return default;
    }
}

File: CodingSample/Java/Advanced/Networking/TCPClientTest/demo1/src/main/java/app/models/ShopModel.java
package app.models;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;

public class ShopModel {
    
    private String server;

    public ShopModel(String server) {
        this.server = server;
    }

    public ItemInfo readItemInfo(String name) throws IOException {
        //Step 1
        Socket connection = new Socket(server, 4000);
        try{
            //Step 2
            var receiver = connection.getInputStream();
            var reader = new BufferedReader(new InputStreamReader(receiver));
            var sender = connection.getOutputStream();
            var writer = new PrintWriter(sender, true);
            reader.readLine(); //read greeting message
            writer.println(name);
            String message = reader.readLine();
            writer.close();
            reader.close();
            if(message != null)
                return ItemInfo.parse(message);
            return new ItemInfo(0, 0);
        }finally{
            //Step 3
            connection.close();
        }
    }
    
}
